### PR TITLE
[AKS Fleet] - Add support for UpgradeStrategy, Default Hubless, and NodeImageSelection

### PR DESCRIPTION
**Pull request by Ealianis**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### Related command
`az fleet`
`az fleet updaterun`
`az fleet upgradestrategy`

### General Guidelines

- [x] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [x] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your pull request is merged into main branch, a new pull request will be created to update `src/index.json` automatically.  
You only need to update the version information in file setup.py and historical information in file HISTORY.rst in your PR but do not modify `src/index.json`.

### CODE REVIEW

**Comment by yonzhan** [comment]
AKS Fleet

File: src/fleet/HISTORY.rst [hunk 1]
```diff
@@ -31,3 +31,14 @@ Release History
...
 ++++++
 * Added support for private fleet & MSI.
+
+0.2.5
+++++++
+* Upgrade SDK version to 2023-08-15-preview
+
+0.2.6
+++++++
+* Hubless is now default fleet creation
```
Comment (jim-minter) [inline src/fleet/HISTORY.rst:41]:
```suggestion
* By default, fleets are now created without a hub
```

File: src/fleet/HISTORY.rst [hunk 2]
```diff
@@ -31,3 +31,14 @@ Release History
...
+* Added support for Fleet Upgrade Strategy
+* Added argument for Node Image Seleciton.
```
Comment (jim-minter) [inline src/fleet/HISTORY.rst:43]:
```suggestion
* Added argument for Node Image Selection.
```

File: src/fleet/azext_fleet/_validators.py [hunk 3]
```diff
@@ -24,6 +24,13 @@ def validate_upgrade_type(namespace):
             "--upgrade-type must be set to 'Full' or 'NodeImageOnly'")
 
 
+def validate_node_image_selection(namespace):
+    node_image_selection = namespace.node_image_selection
+    if node_image_selection not in ("Latest", "Consistent"):
+        raise InvalidArgumentValueError(
+            "--node-image-upgrade-type must be set to 'Latest' or 'Consistent'")
```
Comment (jim-minter) [inline src/fleet/azext_fleet/_validators.py:31]:
```suggestion
            "--node-image-selection must be set to 'Latest' or 'Consistent'")
```

File: src/fleet/azext_fleet/custom.py [hunk 4]
```diff
@@ -29,61 +29,74 @@ def create_fleet(cmd,
...
+        )
+        agent_profile = agent_profile_model(
+            subnet_id=agent_subnet_id
+        )
+        fleet_hub_profile = fleet_hub_profile_model(
+            dns_prefix=dns_name_prefix,
+            api_server_access_profile=api_server_access_profile,
+            agent_profile=agent_profile)
+
+        managed_service_identity = fleet_managed_service_identity_model(type="None")
```
Comment (jim-minter) [inline src/fleet/azext_fleet/custom.py:87]: MSI is not dependent on enable-hub.  It is should always be an available option.

File: src/fleet/azext_fleet/custom.py [hunk 5]
```diff
@@ -101,6 +114,7 @@ def update_fleet(cmd,
                  name,
                  enable_managed_identity=None,
                  assign_identity=None,
+                 no_wait=False,
```
Comment (jim-minter) [inline src/fleet/azext_fleet/custom.py:117]: good catch, but also need to update

g.custom_command("update", "update_fleet", supports_no_wait=True)

i.e. adding supports_no_wait there too.

File: src/fleet/azext_fleet/custom.py [hunk 6]
```diff
@@ -198,14 +212,15 @@ def update_fleet_member(cmd,
                         resource_group_name,
                         name,
                         fleet_name,
-                        update_group=None):
+                        update_group=None,
+                        no_wait=False):
```
Comment (jim-minter) [inline src/fleet/azext_fleet/custom.py:216]: Is this correct?  I think patch update fleet member is always sync, so no-wait is not applicable here?

File: src/fleet/azext_fleet/tests/latest/test_fleet_scenario.py [hunk 7]
```diff
@@ -37,7 +37,7 @@ def generate_ssh_keys(cls):
             key_file.write(TEST_SSH_KEY_PUB)
         return pathname.replace('\\', '\\\\')
 
-    @ResourceGroupPreparer(name_prefix='cli-', random_name_length=8, location='westcentralus')
+    @ResourceGroupPreparer(name_prefix='cli-', random_name_length=8, location='eastus2euap')
     def test_fleet(self):
```
Comment (jim-minter) [inline src/fleet/azext_fleet/tests/latest/test_fleet_scenario.py:42]: expecting to see additional test scenarios added in this file, at least for update strategy?

File: src/fleet/azext_fleet/tests/latest/test_validators.py [hunk 8]
```diff
@@ -129,5 +139,75 @@ def test_empty_identity_id(self):
...
+    def test_valid_node_image_selection(self):
+        valid_node_image_selection = "Latest"
+        namespace = NodeImageSelectionNamespace(valid_node_image_selection)
+
+        self.assertIsNone(validators.validate_node_image_selection(namespace))
+
+    def test_invalid_node_image_selection(self):
+        invalid_node_image_selection = "invalid_node_image_selection"
+        namespace = NodeImageSelectionNamespace(invalid_node_image_selection)
+        err = ("--node-image-selection must be set to 'Latest' or 'Consistent'")
```
Comment (jim-minter) [inline src/fleet/azext_fleet/tests/latest/test_validators.py:186]: presumably these tests are currently failing and will pass once the error string is updated?

**Review by jim-minter — CHANGES_REQUESTED** [review]

File: src/fleet/azext_fleet/tests/latest/test_fleet_scenario.py (hunk 7 above)
Comment (Ealianis) [reply src/fleet/azext_fleet/tests/latest/test_fleet_scenario.py:42]: We will need to create a new scenario. I discussed with Pankil and think it best to do in next version, given the need for this to merge ASAP.

File: src/fleet/azext_fleet/custom.py (hunk 5 above)
Comment (Ealianis) [reply src/fleet/azext_fleet/custom.py:117]: Likewise, good catch.

File: src/fleet/azext_fleet/custom.py (hunk 6 above)
Comment (Ealianis) [reply src/fleet/azext_fleet/custom.py:216]: 0815 it became async. The SDK member update operations return a poller.

Comment (Ealianis) [reply src/fleet/azext_fleet/custom.py:216]: https://github.com/Azure/azure-cli-extensions/blob/99f22f3ea687b4afaa97e2da9268dc731cf03b19/src/fleet/azext_fleet/vendored_sdks/v2023_08_15_preview/aio/operations/_fleet_members_operations.py#L594

File: src/fleet/azext_fleet/custom.py (hunk 4 above)
Comment (Ealianis) [reply src/fleet/azext_fleet/custom.py:87]: Updated.

File: src/fleet/azext_fleet/tests/latest/test_validators.py (hunk 8 above)
Comment (Ealianis) [reply src/fleet/azext_fleet/tests/latest/test_validators.py]: The validator has been removed as there is an `arg_type=get_enum_type function now used on the parameters.
```
arg_type=get_enum_type(['Latest', 'Consistent'])
```

**Review by kairu-ms — APPROVED** [review]

**Pull request merged**